src/blockchain/alt_block_tree.cpp

```diff
@@ -46,259 +46,91 @@ bool AltTree::bootstrap(ValidationState& state) {
-    BlockIndex<AltBlock>& index,
-    const std::vector<typename VbkBlock::id_t>& vbkblocks,
-    AltTree& tree,
-    ValidationState& state,
-    const std::function<bool(const typename VbkBlock::id_t&)>&
-        remove_function) {
-  return vbkBlockCheckDuplicatesInChain(
-      index, vbkblocks, tree, state, remove_function);
-}
-
-template <typename Pop>
-bool payloadsCheckDuplicates(BlockIndex<AltBlock>& index,
-                             std::vector<Pop>& payloads,
-                             AltTree& tree,
-                             ValidationState& state,
-                             bool continueOnInvalid = false) {
-  auto remove_payload =
-      [&payloads, &continueOnInvalid](const typename Pop::id_t& pid) -> bool {
-    if (continueOnInvalid) {
-      for (auto it = payloads.begin(); it != payloads.end(); ++it) {
-        if (it->getId() == pid) {
-          payloads.erase(it);
-          break;
-        }
-      }
-      return true;
-    }
-    return false;
-  };
-
-  return payloadsCheckDuplicates<Pop>(
-      index, map_get_id(payloads), tree, state, remove_payload);
-}
-
 template <typename Pop>
 void commitPayloadsIds(BlockIndex<AltBlock>& index,
                        const std::vector<Pop>& pop,
                        PayloadsStorage& storage) {
   auto pids = map_get_id(pop);
-  index.template insertPayloadIds<Pop>(pids);
+  index.template setPayload<Pop>(pids);
 
   auto containing = index.getHash();
   for (const auto& pid : pids) {
     storage.addAltPayloadIndex(containing, pid.asVector());
   }
 }
 
-// check for Payload duplicates in ancestor and descendant blocks
-template <typename P>
-bool checkNoPayloadDuplicatesInOtherBlocks(
-    AltTree& tree,
-    AltTree::index_t& index,
-    const std::vector<typename P::id_t>& payloads_ids,
-    ValidationState& state) {
-  auto& storage = tree.getStorage();
-
-  for (size_t i = 0; i < payloads_ids.size(); ++i) {
-    auto pid = payloads_ids[i];
-
-    for (const auto& containingBlock :
-         storage.getContainingAltBlocks(pid.asVector())) {
-      auto* containingIndex = tree.getBlockIndex(containingBlock);
-      if (containingIndex == nullptr) {
-        return state.Invalid(
-            "ALT-duplicate-payloads-" + P::name() + "-ancestor",
-            fmt::sprintf(
-                "Ancestor block does not exist in the chain %s that we "
-                "attempted to add to block=%s.",
-                P::name(),
-                index.toPrettyString()));
+template <typename Element, typename Container>
+auto find_duplicates(BlockIndex<AltBlock>& index, Container& pop, AltTree& tree)
+    -> decltype(pop.end()) {
+  const auto startHeight = tree.getParams().getBootstrapBlock().height;
+  Chain<BlockIndex<AltBlock>> chain(startHeight, &index);
+  std::unordered_set<std::vector<uint8_t>> ids;
+
+  const auto& storage = tree.getStorage();
+  auto newend = std::remove_if(pop.begin(), pop.end(), [&](const Element& p) {
```

Comment:
Pop is being modified so the function should be name smth like findDuplicatesAndFilterPop